Bluetooth: Add a possibility to disable bondable mode

MariuszSkamra opened the pull request:
This adds a new function to API to enable/disable bonding flag in SMP Pairing Request/Response.

MariuszSkamra committed:
Bluetooth: Add run-time option to disable SMP bondable mode

This adds a function that will disable Bonding flag in
Authentication Requirements flag in SMP Pairing Request/Response.
This is needed for qualification purposes.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: Add a shell command to disable bondable mode

This adds a shell command for qualification purposes to enable/disable
Bonding flag in Authentication Requirements.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -243,6 +243,7 @@ static struct bt_smp_br bt_smp_br_pool[CONFIG_BT_MAX_CONN];
 #endif /* CONFIG_BT_BREDR */
 
 static struct bt_smp bt_smp_pool[CONFIG_BT_MAX_CONN];
+static bool non_bondable;
```
Why the negative form instead of 'bondable'?

jhedberg reviewed (changes requested):
Is this really needed as a runtime API? I'd have thought that a Kconfig would be sufficient.

carlescufi commented:
I agree with @jhedberg. If we could do with a Kconfig option then I'd honestly prefer that too.

jhedberg commented:
@carlescufi for the automated tester that uses a single image a runtime API is by far the most convenient. What I'd do is add a Kconfig to set the default, and then provide this API to change it at runtime if the app really needs it (but clearly document that it should only be needed in very rare circumstances).

jhedberg commented:
And if we're sure that the only valid use case for the runtime API is for testing, then we could even put it behind CONFIG_BT_TESTING

MariuszSkamra committed:
Bluetooth: Add Kconfig option to disable bondable mode

This adds Kconfig option to allow disable bondable mode in compile
time.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: testing: Exclude Mesh related code if BT_MESH not set

This will exclude testing Mesh related code from build if BT_MESH
option in Kconfig is not set.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -2275,6 +2290,14 @@ static u8_t get_auth(u8_t auth)
+	if (bondable) {
+		auth |= BT_SMP_AUTH_BONDING;
+	} else {
+		auth &= ~(BT_SMP_AUTH_BONDING);
```
minor: the () seem unnecessary

MariuszSkamra replied:
done. I just copied the code few lines above
		auth &= ~(BT_SMP_AUTH_MITM);
:)

jhedberg reviewed (approved):
Looks generally good (one small comment but not critical enough for "request changes")

carlescufi commented on include/bluetooth/testing.h:
```diff
@@ -82,6 +84,15 @@ int bt_test_mesh_lpn_group_remove(u16_t *groups, size_t groups_count);
+ *
+ *  @param enable Value allowing/disallowing to be bondable.
+ */
+void bt_test_set_bondable(bool enable);
```
Might as well add this to the public API?

jhedberg commented on include/bluetooth/conn.h:
```diff
@@ -386,6 +386,15 @@ struct bt_conn_cb {
+/** Enable/disable bonding.
+ *
+ *  Set/clear the Bonding flag in the Authentication Requirements of
+ *  SMP Pairing Request/Response data. The flag is enabled by default.
```
The last sentence is not accurate, is it? The default depends on the Kconfig setting. You should also make it clear that for the vast majority of apps calling this shouldn't be needed. What was the end-product use case that made you decide that this is needed in a wider context than just for testing?

jhedberg replied:
@MariuszSkamra I discussed this with @carlescufi on IRC and the consensus was to keep this public. So please just address the two other issues I brought up.

jhedberg commented on subsys/bluetooth/host/smp.c:
```diff
@@ -243,6 +249,7 @@ static struct bt_smp_br bt_smp_br_pool[CONFIG_BT_MAX_CONN];
 #endif /* CONFIG_BT_BREDR */
 
 static struct bt_smp bt_smp_pool[CONFIG_BT_MAX_CONN];
+static bool bondable = !!BT_SMP_AUTH_BONDING_FLAGS;
```
`static bool bondable = IS_ENABLED(CONFIG_BT_BONDABLE);` might be more readable?

carlescufi reviewed (approved):
@MariuszSkamra great, thank you.

carlescufi commented:
@jhedberg are you fine with the API being public?

jhedberg commented:
> @jhedberg are you fine with the API being public?

If there's a good rationale/use-case for it, sure. Aren't the testing APIs also public? (just that they require enabling CONFIG_BT_TESTING first)

MariuszSkamra commented:
@jhedberg I think @carlescufi has right, putting this to public API, because this is not related to testing only. The end-product might want to use it as well.

Updated
